FWCore/Framework/src/global/OutputModuleBase.cc

```diff
@@ -0,0 +1,380 @@
+                        std::set<BranchID>& keptProductsInEvent) {
+
+      ProductSelector::checkForDuplicateKeptBranch(desc,
+                                                   trueBranchIDToKeptBranchDesc);
+
+      switch (desc.branchType()) {
+      case InEvent:
+        {
+          if(desc.produced()) {
+            keptProductsInEvent.insert(desc.originalBranchID());
+          } else {
+            keptProductsInEvent.insert(desc.branchID());
+          }
+          consumes(TypeToGet{desc.unwrappedTypeID(),PRODUCT_TYPE},
+                   InputTag{desc.moduleLabel(),
+                       desc.productInstanceName(),
+                       desc.processName()});
+          break;
+        }
+      case InLumi:
+        {
+          consumes<InLumi>(TypeToGet{desc.unwrappedTypeID(),PRODUCT_TYPE},
+                           InputTag(desc.moduleLabel(),
+                                    desc.productInstanceName(),
+                                    desc.processName()));
+          break;
+        }
+      case InRun:
+        {
+          consumes<InRun>(TypeToGet{desc.unwrappedTypeID(),PRODUCT_TYPE},
+                          InputTag(desc.moduleLabel(),
+                                   desc.productInstanceName(),
+                                   desc.processName()));
+          break;
+        }
+      default:
+        assert(false);
+        break;
+      }
+      // Now put it in the list of selected branches.
+      keptProducts_[desc.branchType()].push_back(&desc);
+    }
+
+    OutputModuleBase::~OutputModuleBase() { }
+    
+    void OutputModuleBase::doPreallocate(PreallocationConfiguration const& iPC) {
+      auto nstreams = iPC.numberOfStreams();
+      selectors_.resize(nstreams);
+      
+      bool seenFirst = false;
+      for(auto& s : selectors_) {
+        if(seenFirst) {
+          detail::configureEventSelector(selectEvents_,
+                                         process_name_,
+                                         getAllTriggerNames(),
+                                         s);
+        } else {
+          seenFirst = true;
+        }
+      }
+    }
+
+    void OutputModuleBase::doBeginJob() {
+      this->beginJob();
+    }
+    
+    void OutputModuleBase::doEndJob() {
+      endJob();
+    }
+    
+    bool OutputModuleBase::prePrefetchSelection(StreamID id, EventPrincipal const& ep, ModuleCallingContext const* mcc) {
+      
+      auto& s = selectors_[id.value()];
+      return wantAllEvents_ or s.wantEvent(ep,mcc);
+    }
+    
+    bool
+    OutputModuleBase::doEvent(EventPrincipal const& ep,
+                              EventSetup const&,
+                              ActivityRegistry* act,
+                              ModuleCallingContext const* mcc) {
+      
+      {
+          EventSignalsSentry sentry(act,mcc);
+          write(ep, mcc);
+      }
+      if(remainingEvents_ > 0) {
+        --remainingEvents_;
+      }
```

Comment:
This isn't thread safe. ```remainingEvents_``` could have changed from >0 to 0 between the two calls. Probably the following is better
```
auto remainingEvents = remainingEvents_.load();
bool keepTrying = remainingEvents > 0;
while (keepTrying){
  auto newValue = remainingEvents -1;
  keepTrying = not remainingEvents_.compare_exchange_strong(remainingEvents, newValue);
  if( keepTrying) { 
      //the exchange failed because the value was changed by another thread.
      // remainingEvents was changed to be the new value of remainingEvents_;
      keepTrying = remainingEvents >0;
  }
}
```